subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformCachingIntegrationTest.groovy

```diff
@@ -1480,12 +1480,13 @@ ${getFileSizerBody(fileValue, 'new File(outputDirectory, ', 'new File(outputDire
 
     String registerFileSizerWithParameterObject(String fileValue) {
         """                 
-            interface FileSizerParameters extends TransformParameters {
-                @Input
-                Number getValue()
-                void setValue(Number value)
-            }
-            abstract class FileSizer implements TransformAction<FileSizerParameters> {
+            abstract class FileSizer implements TransformAction<Parameters> {
+                interface Parameters extends TransformParameters {
+                    @Input
+                    Number getValue()
+                    void setValue(Number value)
+                }
```

Comment:
Asking for a friend: since this here is Groovy, can this be simplified to:

```groovy
                interface Parameters extends TransformParameters {
                    @Input Number value
                }
```

Reply:
I tried and I get the error message
```
Cannot set the property 'value' because the backing field is final.
```

That seems to be a shortcoming of our injection infrastructure.

Reply:
I think it would be worth adding an issue to clear this up. Maybe look at how we can have minimal ceremony property declarations with Kotlin, too.

Reply:
For Kotlin, this already works (see [here](https://github.com/gradle/gradle/pull/8565/files#diff-74186f02835b24d6a4de4bcc6d40ee93R96)):
```kotlin
interface AddGradleApiParameterNames {
    @get:Input
    var publicApiIncludes: List<String>
    @get:Input
    var publicApiExcludes: List<String>
}
```

Reply:
I tried to create an interface like above in Intellij, and `value` is actually a constant on the interface and not a property. So the idiomatic way cannot work in Groovy.

Reply:
Can it work on an abstract class?

Reply:
Abstract classes also don't work, since Groovy does generate a field for an abstract property - actually putting `@Input abstract Number value` has the same bytecode as `@Input Number value`. We can only isolate managed types, so types where Gradle generates all the fields. Therefore, the abstract Groovy class cannot be used with a field.

Reply:
There's a few things we can tidy up here. For Groovy abstract classes, we can either decide to allow these to be isolated or we can improve the error message when you use an abstract Groovy class. Isolating is awkward because the metaclass is mutable but we could deal with this in a similar way to the other mutable state. In the short term I'd just improve the error message and see what people say.

For the things that look like properties on Groovy interfaces, I wouldn't do anything special. This is just how Groovy works. The same syntax defines a property when present on a class and a static final field when present on an interface. Maybe we just nudge people towards Kotlin instead.

For Kotlin types, we can remove some of the ceremony by handling annotations without the `get:` qualifier.